### PR TITLE
Precursor for reserved delegator slots

### PR DESCRIPTION
**Pull request by j-chmielewski**
This is a minimal PR to make sure relevant breaking changes to data structures are in place for 2.0 / condor release. This allows us to avoid introducing breaking changes in the future release that includes the full implementation of the reserved delegator slots functionality.

Changes:
* added `ValidatorBid::reserved_slots` field to track number of reserved slots (defaulted to 0)
* added `add_reservations` and `cancel_reservations` auction contracts with stubbed implementations
* added `struct Reservation` to track reservations
* added `BidKind::Reservation(Box<Reservation>)` and `BidKindTag::Reservation` variants
* added `BidAddr::Reservation` and `BidAddrTag::Reservation` variants
* added `TransactionEntryPoint::{AddReservations, CancelReservations}` variants
* added auction costs for add/cancel reservations ops

### CODE REVIEW

File: resources/test/sse_data_schema.json [hunk 1]
```diff
@@ -3348,6 +3349,12 @@
           "type": "integer",
           "format": "uint64",
           "minimum": 0.0
+        },
+        "whitelist_size": {
```
Comment (EdHastingsCasperAssociation) [inline resources/test/sse_data_schema.json]: we should not call this a whitelist, instead call it a reserve list

Comment (j-chmielewski) [reply resources/test/sse_data_schema.json]: Renamed the field to `reserved_spots`.

File: types/src/system/auction/constants.rs [hunk 2]
```diff
@@ -62,6 +62,8 @@ pub const ARG_ENTRY_POINT: &str = "entry_point";
 pub const ARG_MINIMUM_DELEGATION_AMOUNT: &str = "minimum_delegation_amount";
 /// Named constrant for `maximum_delegation_amount`.
 pub const ARG_MAXIMUM_DELEGATION_AMOUNT: &str = "maximum_delegation_amount";
+/// Named constant for `reserved_spots`.
+pub const ARG_RESERVED_SPOTS: &str = "reserved_spots";
```
Comment (EdHastingsCasperAssociation) [inline types/src/system/auction/constants.rs]: nitpick: can you change "spots" to "slots" in all places?

File: types/src/transaction/transaction_entry_point.rs [hunk 3]
```diff
@@ -198,7 +228,10 @@ impl TransactionEntryPoint {
             | TransactionEntryPoint::Undelegate
             | TransactionEntryPoint::Redelegate
             | TransactionEntryPoint::ActivateBid
-            | TransactionEntryPoint::ChangeBidPublicKey => false,
+            | TransactionEntryPoint::ChangeBidPublicKey
+            // TODO(jck): make sure this is correct
```
Comment (EdHastingsCasperAssociation) [inline types/src/transaction/transaction_entry_point.rs]: Holds epoch is necessary to calculate available balance, so any operation that includes a balance check requires it.

As I understand it, reservations don't involve a balance check and thus Add / Cancel Reservation should not require holds epoch.

File: types/src/transaction/transaction_v1.rs [hunk 4]
```diff
@@ -650,6 +650,12 @@ impl GasLimited for TransactionV1 {
                             TransactionEntryPoint::ChangeBidPublicKey => {
                                 costs.auction_costs().change_bid_public_key
                             }
+                            TransactionEntryPoint::AddReservation => {
+                                todo!()
+                            }
+                            TransactionEntryPoint::CancelReservation => {
+                                todo!()
+                            }
```
Comment (EdHastingsCasperAssociation) [inline types/src/transaction/transaction_v1.rs]: you need to add cost fields to the chainspec auction costs table

you can set the initial value for both to be the same as `redelegate` and we can adjust as necessary.

File: smart_contracts/contracts/client/add-bid/src/main.rs [hunk 5]
```diff
@@ -12,6 +12,7 @@ fn add_bid(
     delegation_rate: auction::DelegationRate,
     minimum_delegation_amount: Option<u64>,
     maximum_delegation_amount: Option<u64>,
+    reserved_spots: Option<u32>,
```
Comment (j-chmielewski) [inline smart_contracts/contracts/client/add-bid/src/main.rs]: @EdHastingsCasperAssociation do you think this should be an Option with default value = 0? I was told there was some discussion on the topic before and there's some nuance to that.

**Comment by j-chmielewski** [comment]
@EdHastingsCasperAssociation I think this should be enough to allow us to continue the implementation of reservations functionality without introducing breaking changes in the future. Since I'm pretty new to the code base I'd appreciate it if you could confirm that - I think that the main idea of how this will work is now clear.  

My main concerns:
* the add reservations entry point takes a Vector of Reservations - with that in mind, are there any extra requirements that I need to satisfy when it comes to the modified enums / structures?
* are `todo!()` markers in places where it's ok to panic / unreachable - or maybe we want to handle this differently?
* i found multiple chainspec.toml files and am unsure which ones exactly should contain my new auction costs

**Comment by j-chmielewski** [comment]
@EdHastingsCasperAssociation Adding the `Reservation` variant to `BidAddr` changes faucet costs, this seems to be the reason for the failing test. Is this acceptable?

File: smart_contracts/contracts/client/add-bid/src/main.rs (hunk 5 above)
Comment (cspramit) [reply smart_contracts/contracts/client/add-bid/src/main.rs]: @EdHastingsCasperAssociation could you please respond and give a final go to this PR?

**Review by mpapierski — APPROVED** [review]
Perhaps you could reuse some existing error code rather than panic on the host side impl. Otherwise LGTM

**Comment by EdHastingsCasperAssociation** [comment]
> @EdHastingsCasperAssociation Adding the `Reservation` variant to `BidAddr` changes faucet costs, this seems to be the reason for the failing test. Is this acceptable?

Whenever a key variant is added, the size of wasms that import the enum increases which increases costs.

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Pull request merged**